prov/rxm/src/rxm_ep.c

```diff
@@ -675,6 +675,116 @@ static ssize_t rxm_rma_iov_init(struct rxm_ep *rxm_ep, void *buf,
 	return sizeof(*rma_iov) + sizeof(*rma_iov->iov) * count;
 }
 
+static inline ssize_t
+rxm_ep_transmit_generic(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn, void *context,
+			uint8_t count, size_t len, uint64_t data, uint64_t flags,
+			uint64_t tag, uint8_t op, uint64_t comp_flags,
+			struct rxm_tx_buf **tx_buf, struct rxm_tx_entry **tx_entry)
+{
+	struct rxm_pkt *pkt;
+	ssize_t ret;
+
+	*tx_buf = (struct rxm_tx_buf *)rxm_buf_get(&rxm_ep->tx_pool);
+	if (OFI_UNLIKELY(!*tx_buf)) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_DATA, "TX queue full!\n");
+		ret = -FI_EAGAIN;
+		goto err1;
+	}
+
+	*tx_entry = rxm_tx_entry_get(&rxm_ep->send_queue);
+	if (OFI_UNLIKELY(!*tx_entry)) {
+		ret = -FI_EAGAIN;
+		goto err2;
+	}
+
+	(*tx_entry)->ep = rxm_ep;
+	(*tx_entry)->count = count;
+	(*tx_entry)->context = context;
+	(*tx_entry)->flags = flags;
+	(*tx_entry)->tx_buf = *tx_buf;
+	(*tx_entry)->comp_flags |= comp_flags | FI_SEND;
+
+	(*tx_buf)->hdr.msg_ep = rxm_conn->msg_ep;
+
+	pkt = &(*tx_buf)->pkt;
+	rxm_pkt_init(pkt);
```

Comment:
rxm_pkt_init does a memset of 0 for the headers.  As a future optimization, you could look at having different tx_pools based on operation type, with the headers pre-initialized.  This would avoid the memset and have some fields pre-set, such as versions and ops.

Reply:
thanks for the hints. the memset() isn't needed at all. rxm_bug_get() does memset().